### ISSUE TITLE
Reset the sequence of table

### ISSUE THREAD

**Issue by xiaoleATRI** [behave/behave-django#36]
I was using fixtures to load data in the test. 
The problem I found is that the id of django "auth_permission" table was not starting from 1. 
Since the fixtures I was loading are using id of "auth_permission" as foreign key and the error I got is that it can not map the fixture id to the "auth_permission" id in the database.
![edc--mdash--ipython--workplace-edc--mdash---bash--mdash--151-times-40](https://user-images.githubusercontent.com/30299672/30126879-44be0f52-92f2-11e7-9eb4-92540688d74d.jpg)
![pgadmin-4](https://user-images.githubusercontent.com/30299672/30126940-6eda12ae-92f2-11e7-99b9-459e9f44e315.jpg)

@jenisys

**Comment by bittner** [comment]
I assume this is more a question than a bug report, right? Kind of, "how do I make my test work?"

Can you provide appropriate code snippets and fixtures to help analyze your problem? The information currently provided leaves to much room for guessing.

A Personal Statement
------------------------

@mixxorz is certainly better qualified as a discussion partner when it comes to fixures, but I would guess your problem is most likely a general Django testing problem, nothing specifically related to *behave-django*.

Personally, I avoid fixtures in general. They are prone to errors and a horrible maintenance nightmare. Instead I prefer to either write management commands or migrations. -- I know this statement doesn't help you with fixing the current issue, but maybe you could try to revise your testing strategies in the long term.

Said that, let me quote the Django docs on "[Fixture loading](https://docs.djangoproject.com/en/1.11/topics/testing/tools/#fixture-loading)":

> Tests are more readable and it’s more maintainable to create objects using the ORM, for example in [TestCase.setUpTestData()](https://docs.djangoproject.com/en/1.11/topics/testing/tools/#django.test.TestCase.setUpTestData), however, you can also use fixtures.

**Comment by xiaoleATRI** [comment]
Hi, Peter:
Thank you so much for the response, I also plan to use factory_boy to refactor my testing in the future. Currently, the issue would be much more like a configuration problem. 

So, Here it is

1. I have a fixture json file called "test_dump_datadic_data_objects_type_N.json" the snippet of which is as following:
`{
        "pk": 2,
        "model": "authcore.user",
        "fields": 
        {
            "user_permissions": 
            [
                22
            ],
        }
    }`
you can see that there is a field called "user_permissions" which referenced "id" field in django "auth_permission" table
2. And I loaded this fixture before_scenario by using following codes in environment.py
` fixtures = ["core/fixtures/test_dump_datadic_data_objects_type_T.json"]
`
`context.fixtures = fixtures`
3. When I run "python manage.py behave features/", it showed me such error:
![image](https://user-images.githubusercontent.com/30299672/30138959-f5414ba0-931f-11e7-911f-f2b0ba2fbdb9.png)
Which means django can't find the id which equals to "22" in "auth_permission" table
4. Then I checked the "auth_permission" table in test db:
![image](https://user-images.githubusercontent.com/30299672/30139001-4167e99e-9320-11e7-83f0-6272803428c0.png)
Obviously, there is no id equal to 22 

What I'm thinking is that every time when I launch the behave testing, the database should be initialized with new auto_increment sequence which starts from 1 instead of random number.

I also tried command "sqlsequencereset", but it doesn't work

**Comment by mixxorz** [comment]
I looked into it and restructured the way fixtures work. Can you try installing this branch?

https://github.com/behave/behave-django/tree/restructure-testcase

**Comment by xiaoleATRI** [comment]
hi, Mitchel:

I have installed this branch and it solved my previous problem. Thank you so much!
However, another issue comes out:
1. Actually I'm using scenario outline to iterate tests running as following:
`Scenario Outline: Create new topics
    Given I load up fixtures
      |fixture                                                 |
      |core/fixtures/test_dump_datadic_data_objects_type_T.json|
    And I open up the browser "<browser>"
    And I visit the EDC website
    And I log in as admin with "root" and "root"
    When I click "Docs"
    Then I should see "Topic List" in the scope
    When I press the button "New Topic"
    Then I should see "Create a New Topic" in the header
    When I fill the form with "Run Test" and "Create a topic"
    And I press the button "Save"
    Then I should see "Run Test"  in the binding
    Then I close the browser

    Examples:
      | browser | hiptest-uid |
      | Firefox |  |
      | Chrome |  |`
2. When I run the testing, the first round got passed while second round failed
![edc--mdash--ipython--workplace-edc--mdash---bash--mdash--173-times-42](https://user-images.githubusercontent.com/30299672/30177611-4098ec66-93bb-11e7-92e4-68bc254ddd40.jpg)
3. I checked db, and the id was not starting from 1
![keepdb-after-testing](https://user-images.githubusercontent.com/30299672/30177678-6c32489a-93bb-11e7-8e0f-6c21288e5395.jpg)
4. I set the breakpoint during first round execution and checked db at the mean time. The id was starting from 1
![database-at-break-point](https://user-images.githubusercontent.com/30299672/30177758-a74c8a9e-93bb-11e7-8c5c-0f93f770b1cf.jpg)
5. And also when I used "--keepdb" command to rerun the tests, the first round can't even get passed.

I'm wondering why the sequence of "auth_permission" can be override after scenario
@mixxorz

**Comment by mixxorz** [comment]
I've pushed up an update where you can enable `reset_sequences` by setting `context.reset_sequences = True`. I'm pretty sure this would fix your issue.

But I have to say, the root cause of all of this is that you're hard coding primary keys in your fixtures. I think this will cause you more headaches in the future so I suggest you get on with migrating your test data to factory boy instead.

Saying that, a quick fix might just be adding the "Can add user" permission to your fixtures, primary key `22` included.

**Comment by xiaoleATRI** [comment]
It completely solved all my issues. 
Yes, you are right. I should not test on the primary key of table.
I'll change to use factory boy later on.
Thanks again!

**Comment by mixxorz** [comment]
Glad to hear it. 😀

**Comment by xiaoleATRI** [comment]
Hi, @mixxorz 

I have a question which may not be related to behave-django

I'm using behave-django to send Request from form to my Django REST API in backend, the requet method is `POST`.  I believe this test framework is using `LiveServerTestCase`. 

The issue is that I can't post request into database, I was blocked out by server with error 500
![edc-form](https://user-images.githubusercontent.com/30299672/30353162-2540c44e-97d9-11e7-8f86-cdeef387bedd.jpg)
![edc-server-error](https://user-images.githubusercontent.com/30299672/30353164-2a22a7ac-97d9-11e7-849c-897fbdd37eec.jpg)

**Comment by mixxorz** [comment]
Does it work if you test it manually? If not, then there's something wrong with your view code. 😅

**Comment by xiaoleATRI** [comment]
Yes, it works when I run server manually. The weird thing is when I test by using test case, the request can't be processed.

**Comment by mixxorz** [comment]
Sorry it's kind of hard to debug without seeing your code. Does your code also work when DEBUG=False?

**Comment by xiaoleATRI** [comment]
Hi, @mixxorz 

Yes, I turned off DEBUG mode.

Actually, I solved the problem. It's still about loading fixtures. I forgot to load another fixture which is referenced when I make POST request. But still appreciate your help

P.S.
Do you have any good idea on debugging behave? I found it's really hard to debug within step. 
I put 
`import ipdb
  ipdb.set_trace()
` 
in one of step implementation function

And when I run the test, it didn't break at the point I set and pop out the ipdb command line
![edc--mdash--ipython--workplace-edc--mdash---bash--mdash--173-times-42](https://user-images.githubusercontent.com/30299672/30396337-27706756-987e-11e7-8467-e11db1d4e95c.jpg)

what I did is that I put `time.sleep(20000)` within step and manipulate browser manually. Then it will  activate the breakpoint so that I can debug step by step

**Comment by mixxorz** [comment]
I'm not sure why it doesn't break for you. But behave automatically captures stdout output. So for you to use ipdb, you need to run behave with `--no-capture`

**Comment by xiaoleATRI** [comment]
Thanks Man ! The ipdb works now within step

**Comment by xiaoleATRI** [comment]
Hi, @mixxorz
Do you know how to create test database during runtime?

What I'm thinking is as following:
1. Pass in `test_db` name through `python mange.py behave -D test_db='name'` command user data option
2. Create test database in `before_all` function

Since I want to do parallel browser testing, different test bases are needed

Thanks

**Comment by mixxorz** [comment]
While django does support parellel testing with the `--parallel` flag, behave-django doesn't yet support it. Sorry. :/

**Comment by bittner** [comment]
*behave* supports tests to be run in parallel via *behave-parallel*. More details in https://github.com/behave/behave/issues/235.

**Comment by xiaoleATRI** [comment]
Thanks man, it's helpful

**Comment by xiaoleATRI** [comment]
Hi, @mixxorz 

I noticed that you deleted the branch `restructure-testcase` which has been merged into master branch
Can you tell me how to enable `reset_sequences` in `environment.py` ?

`context.reset_sequences=True` didn't work any more

Thanks

**Comment by xiaoleATRI** [comment]
Ok, I found out what the problem is 

The package I downloaded from `pip install` is different from repo in master branch on Github

you may need to double check that

Thanks

@mixxorz

**Comment by mixxorz** [comment]
Yes we've just merged some stuff the past week, for now while there's no 0.6.0 release yet, you can use the master branch.
